Modules/admin/admin_controller.php

```diff
@@ -122,13 +142,34 @@ function read_file($file, $lines) {
                     @fclose($fh);
                 }
                 
-                if ($route->subaction == 'getupdatelog') { 
+                if ($route->subaction == 'getupdatelog' && $session['admin']) { 
                     $route->format = "text";
                     ob_start();
                     passthru("cat /home/pi/data/emonpiupdate.log");
                     $result = trim(ob_get_clean());
                 }
                 
+                if ($route->subaction == 'downloadupdatelog' && $session['admin'])
+                {
+                    header("Content-Type: application/octet-stream");
+                    header("Content-Transfer-Encoding: Binary"); 
+                    header("Content-disposition: attachment; filename=emonpiupdate.log");
+                    header("Pragma: no-cache"); 
+                    header("Expires: 0");
+                    flush();
+                    if (file_exists("/home/pi/data/emonpiupdate.log"))
```

Comment:
that would not work as far as i can tell the variables are only known on that route (in that if statement). I can move them outside the if statement and reference adapt my code as well as the other code that was there before to use the variable. The only not so clean down side is that the variables would be set each time you use the emonpi route action while perhaps never having need for them although i don't really consider that a bad thing as it would not have an impact and then you can reference the variables

the code actually used harcoded absolute paths before i made my changes ..